packages/datadog-plugin-moleculer/src/server.js

```diff
@@ -0,0 +1,62 @@
+'use strict'
+
+function createWrapRegisterMiddlewares (tracer, config) {
+  return function wrapRegisterMiddlewares (registerMiddlewares) {
+    return function registerMiddlewaresWithTrace (userMiddlewares) {
+      if (this.middlewares && this.middlewares.add) {
+        this.middlewares.add(createMiddleware(tracer, config))
+      }
+
+      return registerMiddlewares.apply(this, arguments)
+    }
+  }
+}
+
+function createMiddleware (tracer, config) {
+  return {
+    name: 'Datadog',
+
+    localAction (next, action) {
+      return function datadogMiddleware (ctx) {
+        const childOf = tracer.extract('text_map', ctx.meta)
+        const service = ctx.service || {}
+        const action = ctx.action || {}
+        const options = {
+          service: config.service,
+          resource: action.name,
+          type: 'web',
+          tags: {
+            'span.kind': 'server',
+            'moleculer.service': service.name,
+            'moleculer.action': action.name,
+            'moleculer.request_id': ctx.requestID,
+            'moleculer.node_id': ctx.nodeID
+          }
```

Comment:
Would be good, if we can process the `action.tracing` schema and we can add tags from `ctx.params` and `ctx.meta` same as built-in tracer does:
https://github.com/moleculerjs/moleculer/blob/34431aeb634a323159c89aff19c619e9ec8cd438/src/middlewares/tracing.js#L54-L70

Reply:
I think it might be a bit confusing to use Moleculer config from dd-trace as users would expect all their tracing configuration to be centralized. I can definitely add equivalent options to dd-trace however to get the same functionality.

Reply:
For now I added options as booleans only since `params` can be any type so it would be difficult to filter. This is especially true since every service and every action might have different params, so a global filter may not be that helpful. Maybe a blocklist would make more sense instead, for example `['password']` if it's a known security risk for an app. Let me know what you think but in any case it can always be improved in a future PR.

Reply:
Great. Where should I set `params: true`? My init looks like:
![image](https://user-images.githubusercontent.com/306521/131252514-6ba788ce-3418-4474-906d-6c78404821a5.png)

Reply:
```js
tracer.use('moleculer', {
  params: true
})
```

You also don't need the additional `use` calls for anything that doesn't need configuration since all plugins are enabled by default.

Reply:
I don't see the params:
![image](https://user-images.githubusercontent.com/306521/131252891-50aa506e-350f-4d38-9316-8c2499ada6c4.png)

![image](https://user-images.githubusercontent.com/306521/131252898-7c46936c-3f61-4d6f-94f5-c70d65e5ee28.png)

Using the f135d1c commit.

Reply:
Are you testing this with one of the examples from the repo that I could try locally?

Reply:
I've made the following steps for test:
1. Generate a Molecule sample project with `moleculer init project mol-dd-trace`
2. Install this PR branch with `npm i DataDog/dd-trace-js#moleculer`
3. Add the above codes to the top of `moleculer.config.js` file
4. `npm run dev`
5. Open http://localhost:3000 and click "Greeter" and "Welcome" action which has a `name` params.

Reply:
I didn't realize that Molecular uses a CLI to run the app. Since dd-trace must be initialized before Moleculer is imported for things to work properly, it means that the initialization must happen in a separate file. When not needing plugin configuration, this can be achieved with simply `NODE_OPTIONS='-r dd-trace/init'`. When needing plugin configuration or other programmatic configuration, then you'd have to move dd-trace usage to a separate file. For example:

```js
// ./datadog.js

const tracer = require("dd-trace").init();

tracer.use("moleculer", {
  params: true
});
```

And then import this file from the CLI:

```
NODE_OPTIONS='-r ./datadog' npm run dev
```

Reply:
I've tried but now I've got it:
![image](https://user-images.githubusercontent.com/306521/131705382-0511aa1b-4ead-4544-85ea-2a2232adcd97.png)

Reply:
Is this still in the same simple app? Not sure how it could run out of memory unless it's in a very constrained environment. Is this something I can reproduce? Was this run on your machine directly?

Reply:
Yeah, the same project what I described above. Just now I executed what you suggested with "-r datadog"

Reply:
Not sure if it's helpful in this case, but I've seen a heap out of memory error on several different occasions when the tracer tries to add a large error as a span tag. In my case it's usually an axios error that contains nested circular references to the request and response objects.

Reply:
@icebob Can you provide more detailed reproduction steps? I tried sending tons of requests to the Greeter service and I never run out of memory.

Reply:
Here is my test project: https://github.com/icebob/mol-dd-trace-repro